tests/Unit/Evolution/Systems/Cce/Test_KleinGordonWorldtubeData.cpp

```diff
@@ -0,0 +1,362 @@
+// is used to read the worldtube data and verify that the loaded data matches
+// the originally generated data.
+//
+// The test involves three buffer updaters
+// (1) buffer_updater: a `KleinGordonWorldtubeH5BufferUpdater` object, built
+//     from a HDF5 file to be written.
+// (2) serialized_and_deserialized_updater: serialized-deserialized
+//     buffer_updater.
+// (3) dummy_buffer_updater: an object of `KleinGordonDummyBufferUpdater`
+//     defined above.
+//
+// In the test, we treat `dummy_buffer_updater` as a reference object, whose
+// stored modal data are to be compared with. For `buffer_updater` and
+// `serialized_and_deserialized_updater`, we check:
+//
+// (1) The extraction radius is correctly retrieved.
+// (2) The time stamps are the same as the ones we generate.
+// (3) The worldtube data are the same as `dummy_buffer_updater`.
+template <typename Generator>
+void test_klein_gordon_worldtube_buffer_updater(
+    const gsl::not_null<Generator*> gen,
+    const bool extraction_radius_in_filename) {
+  UniformCustomDistribution<double> value_dist{0.1, 0.5};
+
+  const double extraction_radius = 100.0;
+
+  const double frequency = 0.1 * value_dist(*gen);
+  const double amplitude = 0.1 * value_dist(*gen);
+  const double target_time = 50.0 * value_dist(*gen);
+
+  const size_t buffer_size = 4;
+  const size_t interpolator_length = 3;
+  const size_t file_l_max = 8;
+  const size_t computation_l_max = 10;
+
+  const std::string filename = extraction_radius_in_filename
+                                   ? "BoundaryDataH5Test_CceR0100.h5"
+                                   : "BoundaryDataH5Test.h5";
+  if (file_system::check_if_file_exists(filename)) {
+    file_system::rm(filename, true);
+  }
+
+  // Generate fake modal data and write to `filename`
+  {
+    // scoped to close the file
+    Scalar<ComplexModalVector> kg_psi_coefficients;
+    Scalar<ComplexModalVector> kg_pi_coefficients;
+
+    Cce::ReducedWorldtubeModeRecorder recorder{filename};
+    for (size_t t = 0; t < 20; ++t) {
+      const double time = 0.01 * static_cast<double>(t) + target_time - 0.1;
+
+      TestHelpers::create_fake_time_varying_klein_gordon_modal_data(
+          make_not_null(&kg_psi_coefficients),
+          make_not_null(&kg_pi_coefficients), extraction_radius, amplitude,
+          frequency, time, file_l_max);
+
+      recorder.append_worldtube_mode_data(
+          "/KGPsi", time, get(kg_psi_coefficients), file_l_max, true);
+      recorder.append_worldtube_mode_data(
+          "/dtKGPsi", time, get(kg_pi_coefficients), file_l_max, true);
+    }
+  }
+
+  // Create a `KleinGordonWorldtubeH5BufferUpdater` object `buffer_updater`
+  // from the HDF5 file `filename` written above.
+  // Then examine its extraction radius and time stamps.
+  auto buffer_updater =
+      extraction_radius_in_filename
+          ? KleinGordonWorldtubeH5BufferUpdater{filename}
+          : KleinGordonWorldtubeH5BufferUpdater{filename, extraction_radius};
```

Comment:
Is it important to do a test where the extraction radius isn't in the filename AND you also pass a `std::nullopt` to the constructor for the extraction radius?

Reply:
I feel it's not, because the radius is never used by both Klein-Gordon and Bondi.